tools/optimizer-server/src/main.rs

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::{env, fs, io, io::Write, os::unix::io::AsRawFd, path::Path};
+
+use nix::{
+    poll::{poll, PollFd, PollFlags},
+    sched::{setns, CloneFlags},
+    unistd::{
+        fork,
+        ForkResult::{Child, Parent},
+    },
+};
+
+use fanotify::low_level::*;
+
+const DEFAULT_TARGET: &str = "/";
+
+#[derive(Debug)]
+pub enum SetnsError {
+    IO(io::Error),
+    Nix(nix::Error),
+}
+
+fn get_pid() -> Option<String> {
+    match env::var("_MNTNS_PID") {
+        Ok(string) => Some(string),
+        Err(_) => None,
+    }
+}
+
+fn get_target() -> String {
+    match env::var("_TARGET") {
+        Ok(string) => string,
+        Err(_) => DEFAULT_TARGET.to_string(),
+    }
+}
+
+fn get_fd_path(fd: i32) -> Option<String> {
+    let fd_path = format!("/proc/self/fd/{}", fd);
+    if let Ok(path) = fs::read_link(fd_path) {
+        Some(path.to_string_lossy().to_string())
+    } else {
+        None
+    }
+}
+
+fn set_ns(ns_path: String, flags: CloneFlags) -> Result<(), SetnsError> {
+    fs::File::open(Path::new(ns_path.as_str()))
+        .map(|file| {
+            setns(file.as_raw_fd(), flags)
+                .map_err(|e| SetnsError::Nix(e))
+                .unwrap()
+        })
+        .map_err(|e| SetnsError::IO(e))
+}
+
+fn init_fanotify() -> Result<i32, io::Error> {
+    fanotify_init(
+        FAN_CLOEXEC | FAN_CLASS_CONTENT | FAN_NONBLOCK,
+        O_RDONLY | O_LARGEFILE,
+    )
+}
+
+fn mark_fanotify(fd: i32, path: &str) -> Result<(), io::Error> {
+    fanotify_mark(
+        fd,
+        FAN_MARK_ADD | FAN_MARK_MOUNT,
+        FAN_OPEN | FAN_CLOSE_WRITE,
+        AT_FDCWD,
+        path,
+    )
+}
+
+fn send_path(writer: &mut io::Stdout, path: String) -> Result<(), io::Error> {
+    writer.write_all(path.as_bytes())?;
+    writer.flush()
+}
+
+fn handle_fanotify_event(fd: i32) {
+    let mut writer = io::stdout();
+
+    let mut fds = [PollFd::new(fd.as_raw_fd(), PollFlags::POLLIN)];
+    loop {
+        let poll_num = poll(&mut fds, -1).unwrap();
```

Comment:
Try to avoid directly `unwrap` unless we be sure that it's safe

Reply:
Fixed it with `match`.